### PR TITLE
Working Hub can log methods and store data in Database

### PR DESCRIPTION
**Pull request by OlsonBein**

### CODE REVIEW

File: SmartHouse.Hub/SmartHouse.Hub/Hubs/SmartHouseHub.cs [hunk 1]
```diff
@@ -47,74 +64,74 @@ public override async Task OnDisconnectedAsync(Exception ex)
 
         public async Task<bool> InitSensor(Sensor newSlave)
```
Comment (KarachentsevDanil) [inline SmartHouse.Hub/SmartHouse.Hub/Hubs/SmartHouseHub.cs]: Add cancellation token to method

File: SmartHouse.Hub/SmartHouse.Hub/Hubs/SmartHouseHub.cs [hunk 2]
```diff
@@ -47,74 +64,74 @@ public override async Task OnDisconnectedAsync(Exception ex)
...
-            if (newSlave != null)
+            try
             {
-                bool success = _slaves.TryAdd(Context.ConnectionId, newSlave);
-                return success;
+                if (newSlave != null)
+                {
+                    bool success = _slaves.TryAdd(Context.ConnectionId, newSlave);
+                    await _sensorRepository.Add(newSlave);
+                    await Clients.Caller.SendAsync("GetUrlForListen", _config.UrlForListen);
```
Comment (KarachentsevDanil) [inline SmartHouse.Hub/SmartHouse.Hub/Hubs/SmartHouseHub.cs]: Move "GetUrlForListen" to constant , we should avoid magic string

File: SmartHouse.Hub/SmartHouse.Hub/Hubs/SmartHouseHub.cs [hunk 3]
```diff
@@ -47,74 +64,74 @@ public override async Task OnDisconnectedAsync(Exception ex)
...
+            }
+            catch (Exception ex)
+            {
+                Logger.Error(ex, $"Something wrong with {newSlave.MAC}");
             }
-
             return false;
         }
 
         public async Task<bool> InitDevice(Device newSlave)
```
Comment (KarachentsevDanil) [inline SmartHouse.Hub/SmartHouse.Hub/Hubs/SmartHouseHub.cs]: Add cancellation token to all async methods

File: SmartHouse.Hub/SmartHouse.Hub/Hubs/SmartHouseHub.cs [hunk 4]
```diff
@@ -47,74 +64,74 @@ public override async Task OnDisconnectedAsync(Exception ex)
...
+                if (newSlave != null)
                 {
-                    await Clients.Caller.SendAsync(on.Adress, on.Args);
+                    newSlave.ConnectionId = Context.ConnectionId;
+                    foreach (var item in newSlave.Methods)
+                    {
+                        item.ConnectionId= Context.ConnectionId;
+                    }
+                    bool success = _slaves.TryAdd(newSlave.MAC, newSlave);
+                    // пробежать по всем методам и заменить слейва на текущий ид (Context.ConnectionId)
```
Comment (KarachentsevDanil) [inline SmartHouse.Hub/SmartHouse.Hub/Hubs/SmartHouseHub.cs]: We should use English for communication!

File: SmartHouse.Hub/SmartHouse.Hub/Model/UrlDataConfiguration.cs [hunk 5]
```diff
@@ -7,7 +7,7 @@ public class UrlDataConfiguration
     {
         public int PortToListen { get; set; }
```
Comment (KarachentsevDanil) [inline SmartHouse.Hub/SmartHouse.Hub/Model/UrlDataConfiguration.cs]: Ad empty line between all properties

File: SmartHouse.Hub/SmartHouse.Hub/Startup.cs [hunk 6]
```diff
@@ -19,23 +23,23 @@ public Startup(IHostingEnvironment env)
...
         }
 
-
         public void ConfigureServices(IServiceCollection services)
         {
             services.Configure<UrlDataConfiguration>((settings) =>
                Configuration.GetSection("UrlConfiguration").Bind(settings));
+            services.Configure<DbConnectionDataConfiguration>((settings) =>
+               Configuration.GetSection("DbConfiguration").Bind(settings));
+            services.AddSingleton<IRepository<Device>>(s => new Repository<Device>(new Database("mongodb://localhost:27017/smartHub")));
```
Comment (KarachentsevDanil) [inline SmartHouse.Hub/SmartHouse.Hub/Startup.cs]: Move mongo connection string to settings

File: SmartHouse.Hub/SmartHouse.Hub/Hubs/SmartHouseHub.cs [hunk 7]
```diff
@@ -47,74 +64,74 @@ public override async Task OnDisconnectedAsync(Exception ex)
...
+            catch (Exception ex)
+            {
+                Logger.Error(ex, $"Something wrong with {newSlave.MAC}");
             }
+            return false;
+        }
 
+        public async Task<bool> Manage(Device newSlave)
+        {
+            HouseSlaveInvoker on = newSlave.Methods.Where(x => x.Name.Contains("On")).FirstOrDefault();
```
Comment (Artem-Marakhovskyi) [inline SmartHouse.Hub/SmartHouse.Hub/Hubs/SmartHouseHub.cs]: Strings "On" and "Off" should be consts but not literals

Comment (Artem-Marakhovskyi) [inline SmartHouse.Hub/SmartHouse.Hub/Hubs/SmartHouseHub.cs]: Use `Methods.FirstOfDefault(x => x.Name.Equals("On"))`

File: SmartHouse.Hub/SmartHouse.Hub/Hubs/SmartHouseHub.cs [hunk 8]
```diff
@@ -47,74 +64,74 @@ public override async Task OnDisconnectedAsync(Exception ex)
...
+            HouseSlaveInvoker off = newSlave.Methods.Where(x => x.Name.Contains("Off")).FirstOrDefault();
```
Comment (Artem-Marakhovskyi) [inline SmartHouse.Hub/SmartHouse.Hub/Hubs/SmartHouseHub.cs]: This method has 2 responsibilities.

File: SmartHouse.Hub/SmartHouse.Hub/Hubs/SmartHouseHub.cs (hunk 7 above)
Comment (Artem-Marakhovskyi) [reply SmartHouse.Hub/SmartHouse.Hub/Hubs/SmartHouseHub.cs]: Method name Manage is too generalized.

File: SmartHouse.Hub/SmartHouse.Hub/UdpListener.cs [hunk 9]
```diff
@@ -14,6 +14,9 @@ public class UdpListener
     {
         private UrlDataConfiguration _config;
 
+        public UdpListener()
```
Comment (Artem-Marakhovskyi) [inline SmartHouse.Hub/SmartHouse.Hub/UdpListener.cs]: Remove empty ctor

File: SmartHouse.Hub/SmartHouse.Hub/Hubs/SmartHouseHub.cs [hunk 10]
```diff
@@ -1,56 +1,92 @@
-﻿
-using Microsoft.AspNetCore.SignalR;
+﻿using Microsoft.AspNetCore.SignalR;
```
Comment (Artem-Marakhovskyi) [inline SmartHouse.Hub/SmartHouse.Hub/Hubs/SmartHouseHub.cs]: Tip: the class is huge with a lot of funcitonality. In such cases delegate responsibility to other classes and avoid making it a snowball

**Pull request merged**